### ISSUE TITLE
JenkinsPipelineUnit tests + JaCoCo

### ISSUE THREAD

**Issue by Moser-ss** [jenkinsci/JenkinsPipelineUnit#103]
Hey, I am thinking to add JaCoCo plugin to the JenkinsPipelineUnit tests, do you think that it is possible? Or with will work to have code coverage in the Jenkins pipelines?

**Comment by mkobit** [comment]
Getting coverage over how Jenkins actually runs pipelines is pretty damn difficult (see https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin/issues/32 for my attempt). Are you asking about adding coverage to the _JenkinsPipelineUnit _ tests, or using it in your own project?

**Comment by hypery2k** [comment]
I'm also interested. Would like to have coverage of my unit tests for the jenkins pipeline library

**Comment by vickychijwani** [comment]
Same here. We're building a framework for other developers in the company and it would be great to have coverage reports to guide testing.

**Comment by idij** [comment]
Not JaCoCo, but I have code coverage with Cobertura working for my Jenkins shared library Spock tests. (I'm using @macg33zr's [pipelineUnit](https://github.com/macg33zr/pipelineUnit) approach in general). I had to modify the loader though.  See #119 for the outline of the changes needed to JenkinePipelineUnit. Possibly the same approach can be used for JaCoCo.

I also have it working with an unmodified JenkinsPipelineUnit by extending the needed classes and injecting an extended helper with the changes from #119 + cut & paste of the bits of the original classes that can't be overridden.

**Comment by haridsv** [comment]
I found that if I moved the code from the var file into a class that extends `CpsScript` I could then get the coverage without any extra effort. E.g., take this var file `vars/hello.groovy`

```
void call(String name) {
    echo "Hello, ${name}!"
}
```

The code can be moved to a class like this:

```
package vars

import org.jenkinsci.plugins.workflow.cps.CpsScript

abstract class Hello extends CpsScript {
    void call(String name) {
        echo "Hello, ${name}!"
    }
}
```

The var file itself will simply change to one liner like this:

```
@groovy.transform.BaseScript vars.Hello steps
```

There is no impact of this refactoring on any of the client code but now JaCoCo code coverage will work right out of the box. However, there are no other immediate advantages that we can expect out of this, e.g., the fields in this class would still need to be static as you will end up with multiple instances of it (just like for var files). Also, you would still load the var file as a script for testing as the new class is an abstract class.

There may be other benefits that I haven't had chance to explore. E.g., it is now possible to have one var class extend another or have a common base class to create a framework.

**Comment by cmur2** [comment]
@haridsv can you maybe share more of your approach?
I tried to replicate it but (using Gradle) get a failure on task `:compileGroovy` with an exception that is caused by `java.lang.NoSuchFieldError: BINDING_TYPE`. This is using Groovy 2.4.10.

**Comment by mlasevich** [comment]
> @haridsv can you maybe share more of your approach?
> I tried to replicate it but (using Gradle) get a failure on task `:compileGroovy` with an exception that is caused by `java.lang.NoSuchFieldError: BINDING_TYPE`. This is using Groovy 2.4.10.

I am getting a whole other set of problems with this approach - seems like you are not allowed to extend CpsScript due to it throwing exception from constructor (and Groovy does not seem to deal with this)

I did find a simple workaround that seems to work. Instead of extending CpsScript explicitly, I just created an empty groovy file (MyScriptBase.groovy) and extended from that. That seemed to work.

**Comment by mamuf** [comment]
JaCoCo actually worked to a satisfaction up to version 1.3, but it broke somewhere between 1.3 and 1.4. Since 1.4 JaCoCo reports way less coverage than before (around 28 % instead of 77 % in our case). It would be great if this could be solved since the new versions provide very useful features.

**Comment by malashi** [comment]
@mamuf  can you please share your steps?
Also, for all in this thread, are there any progress/plans to have coverage enabled? Thanks.

**Comment by goober** [comment]
For those who might stumble upon this issue.
I got it working in an ok manner by tweaking my build.gradle file as described in https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin/issues/32#issuecomment-765397659

**Comment by dblock** [comment]
Specifically, the workaround is to add

```groovy
test {
    jacoco {
        classDumpDir = file("$buildDir/jacoco/classpathdumps")
    }
}
```

and

```
jacocoTestReport {
    afterEvaluate {
        classDirectories.from = fileTree("$buildDir/jacoco/classpathdumps") {
            include '*'
        }
    }
}
```

See https://github.com/opensearch-project/opensearch-build/pull/980

**Comment by deporcali** [comment]
I tried using the `classDumpDir` workaround mentioned by @dblock in [this comment](https://github.com/jenkinsci/JenkinsPipelineUnit/issues/103#issuecomment-971588611) but these are the limitations that I found

1. Only loaded classes will be dumped (i.e. classes not executed at all will not show-up in your report as not covered)
2. Jacoco agent will dump all third library classes as well, and hardcoding the "includes/excludes" is not efficient as new files with new naming conventions can be created
3. Report generation fails with `Can't add different class with same name` since this Jenkins unit testing framework will create several versions of the class at runtime. Jacoco dumps these class files with duplicate classes and therefore it throws this error (see example below)
```bash
# Example source files
vars/foo.groovy
src/mypackage/bar.groovy

# Example class files dumped by Jacoco
classdumpdir/foo.7bc27aq92d03dda5.class
classdumpdir/foo.aea79dlf0a9e3a59.class
classdumpdir/foo.b0baq2fdb441c0f6.class
classdumpdir/mypackage/bar.b0bfqdpdb441c0f6.class
```

Now knowing all these limitations, the best approach was to

1. Create the Jacoco xml report with the regular classes compiled (not the dumped ones by the Jacoco agent) so the report will include all classes even if they are not covered
2. Since there is not a way to "merge" the duplicated classes in order to avoid the `Can't add different class with same name` error, then the best approach is to separate all the duplicated classes into separate directories and then create individual xml Jacoco reports for each directory
3. Merge all the Jacoco xml reports for a final code coverage analysis. For this step I am using SonarQube since there is an option to pass in several xml reports through the use of `sonar.coverage.jacoco.xmlReportPaths` (if you are not using SonarQube then this approach might not work for you, or you will have to do some research in order to see if there is a way to merge the xml files locally)

With that being said this is the code that you can just copy/paste (tested with Gradle 4.6) and it will automatically exclude the necessary class files from your project's source files configuration, will do all the parsing necessary in order to avoid duplicated classes at analysis time, and will create several Jacoco xml reports under `build/reports/jacoco/xml/*.xml` so you can feed them to any other code coverage analysis tool

```groovy
import java.nio.file.Files
import java.nio.file.Paths

apply plugin: 'groovy'
apply plugin: 'jacoco'

jacoco {
    toolVersion = "0.8.5"
}

def classPathDumpsDir = "$buildDir/jacoco/classpathdumps"
def parsedClassPathDumpsDir = "$buildDir/jacoco/parsedclasspathdumps"

sourceSets {
    main {
        groovy {
            srcDirs = ['src', 'vars']
        }
    }
}

test {
    jacoco {
        classDumpDir = file(classPathDumpsDir)
    }
}

jacocoTestReport {
    reports {
        xml.enabled true
        html.enabled false
        xml.destination "${buildDir}/reports/jacoco/xml/${System.getProperty('coverageFileName', 'jacoco.xml')}"
    }
    afterEvaluate {
        classDirectories = fileTree("${System.getProperty('coverageClassesDir', sourceSets.main.groovy.outputDir.path)}")
    }
}

task jacocoParseClassDumps {
    doFirst {
        cleanClassPathDumps(classPathDumpsDir)
        parseClassPathDumps(classPathDumpsDir, parsedClassPathDumpsDir)
    }
}

task jacocoTestReports {
    def classesDirs = []

    classesDirs << "${sourceSets.main.groovy.outputDir}"

    if (file(parsedClassPathDumpsDir).exists()) {
        file(parsedClassPathDumpsDir).eachFile {
            classesDirs << it.path
        }
    }

    dependsOn classesDirs.collect { "runJacocoWith${file(it).name}" }

    classesDirs.each {
        def targetDir = file(it)
        def targetDirName = targetDir.name
        def targetDirPath = targetDir.path
        def targetFileName = targetDirName + '.xml'
        task "runJacocoWith${targetDirName}"(type: GradleBuild) {
            buildFile = 'build.gradle'
            tasks = ['jacocoTestReport']
            startParameter.systemPropertiesArgs += [coverageFileName: targetFileName, coverageClassesDir: targetDirPath]
        }
    }
}

// --------------------------------------------------
// Helper functions

/**
 * Collects all the source file paths from all the source folders specified in the project configuration.
 *
 * @return A list of all the source file paths relative to the root package and excluding the
 * extensions (i.e. [foo, mypackage/bar])
 */
def getAllSourceFilePaths() {
    def sourceFilePaths = []
    def baseDirs = sourceSets.main.groovy.srcDirs
    baseDirs.each { File baseDir ->
        fileTree(dir: baseDir).each { File file ->
            sourceFilePaths << file.path.replace(baseDir.path + '/', '').replace('.groovy', '')
        }
    }

    return sourceFilePaths
}

/**
 * Cleans the class path dump directory by deleting class files that were not compiled from the source files (i.e. class
 * files from external libraries, etc).
 *
 * @param classPathDumpsDir Path to the class dumps directory seen by the Jacoco agent.
 */
def cleanClassPathDumps(String classPathDumpsDir) {
    def srcFilePaths = getAllSourceFilePaths()
    fileTree(dir: classPathDumpsDir).each { File file ->
        def filePath = file.path.replace(classPathDumpsDir + '/', '')
        def cleanedFilePath = filePath.contains('$') ? filePath.tokenize('$')[0] : filePath.tokenize('.')[0]
        if (!srcFilePaths.contains(cleanedFilePath) || file.path.contains('$')) {
            file.delete()
        }
    }
}

/**
 * Check whether the class is already duplicated in the specified directory.
 *
 * @param baseDir Path to the directory containing the classes to check.
 * @param relativeFilePath Relative path of the class file to check for duplication.
 *
 * @return True if there is already a class file with the same class as the one we are checking, or False otherwise.
 */
def isClassDuplicatedInDir(String baseDir, String relativeFilePath) {
    for (File file in fileTree(dir: baseDir)) {
        def cleanRelativeFilePath = relativeFilePath.contains('$') ? relativeFilePath.tokenize('$')[0] : relativeFilePath.tokenize('.')[0]
        if (file.path.contains("${baseDir}/${cleanRelativeFilePath}")) {
            return true
        }
    }
    return false
}

/**
 * Parses the class dump directory seen by the Jacoco agent by copying duplicated file names to individual folders. If
 * this is not done, then at the time of analysis it will throw 'Can't add different class with same name' error since
 * the Jenkins unit testing framework dynamically creates different class files for the same class at runtime.
 *
 * @param classPathDumpsDir Path to the class files dump directory from the Jacoco agent.
 * @param parsedClassPathDumpsDir Path to the base directory that will contain the sub-directories containing the parsed
 * class files.
 */
def parseClassPathDumps(String classPathDumpsDir, String parsedClassPathDumpsDir) {
    def classesBaseTree = fileTree(dir: classPathDumpsDir)
    for (File srcFile in classesBaseTree) {
        for (int i = 1 ; i <= classesBaseTree.size() ; i++) {
            def targetDir = "${parsedClassPathDumpsDir}/${i}"
            if (Files.notExists(Paths.get(targetDir))) {
                Files.createDirectories(Paths.get(targetDir))
            }

            def relativeTargetFilePath = srcFile.path.replace(classPathDumpsDir + '/', '')
            def targetFile = file("${targetDir}/${relativeTargetFilePath}")

            if (!isClassDuplicatedInDir(targetDir, relativeTargetFilePath)) {
                def targetParentPath = Paths.get(targetFile.parent)
                if (Files.notExists(targetParentPath)) {
                    Files.createDirectories(targetParentPath)
                }
                Files.copy(Paths.get(srcFile.path), Paths.get(targetFile.path))
                break
            }
        }
    }
}
```

So with this code if we run the following commands (note the two separate gradle calls are a must, they cannot be combined into one due to how Gradle configures the dynamic tasks)

```bash
./gradlew test jacocoParseClassDumps
./gradlew jacocoTestReports
```

Then it will dump and parse the class files, and also create the corresponding xml Jacoco reports

```bash
# Example source files
vars/foo.groovy
src/mypackage/bar.groovy

# Example parsed class files dumped by Jacoco
build/jacoco/parsedclasspathdumps/1/mypackage/bar.b0bfqdpdb441c0f6.class
build/jacoco/parsedclasspathdumps/1/foo.7bc27aq92d03dda5.class
build/jacoco/parsedclasspathdumps/2/foo.aea79dlf0a9e3a59.class
build/jacoco/parsedclasspathdumps/3/foo.b0baq2fdb441c0f6.class

# Example xml Jacoco reports generated
build/reports/jacoco/xml/main.xml
build/reports/jacoco/xml/1.xml
build/reports/jacoco/xml/2.xml
build/reports/jacoco/xml/3.xml
```

**Comment by dblock** [comment]
@deporcali This was no small feat! What's the "right" way to fix this?

**Comment by mahmoud-ashi** [comment]
@dblock I actually used your solution from https://github.com/jenkinsci/JenkinsPipelineUnit/issues/103#issuecomment-971588611 and adapted it a bit to handle the issue that @deporcali mentioned. It's a bit "hacky" but works nicely and generates a single jacoco.xml file that I then use with the jacoco Jenkins plugin to show it on the build's page. Here is my solution which works when I have the following sources:

- vars/*.groovy
- src/com/company/**/*.groovy

```
test {
    jacoco {
        classDumpDir = file("$buildDir/jacoco/classpathdumps")
    }

    finalizedBy jacocoTestReport
}

jacocoTestReport {
    dependsOn test

    afterEvaluate {
        /*
         * There are 2 location where the ".class" files exist:
         * 1) "$buildDir/jacoco/classpathdumps": This is where jacoco puts the class files which have been triggered by
         *    the tests.
         * 2) "$buildDir/classes/groovy/main": This is where gradle places all the class files during compilation
         *
         * It should be enough to just include the file tree from (1) above. However, this will result in not
         * considering any of the files which has not been tested.
         * In order to consider those as well, we add the class files from (2) as well. However, if we try to add the
         * same class multiple times, jacoco will complain. Therefore, we need to add all the files from (1), and the
         * files from (2) which are not in (1) already.
         */
        String jacocoClassesPath = "$buildDir/jacoco/classpathdumps"
        String gradleClassesPath = "$buildDir/classes/groovy/main"
        // The single "*" here indicates the "default" package which in our case is the "vars" dir
        List includesList = ["*", "com/company/**"]
        List excludesList = ["**/*Test*", "**/*Mock*"]

        List allTrees = [
            fileTree(dir: jacocoClassesPath, includes: includesList, excludes: excludesList),
            fileTree(dir: gradleClassesPath, includes: includesList, excludes: excludesList).filter { f ->
                /*
                 * In this filter, we want to remove all the files which were already covered by jacoco in the first
                 * `fileTree` above. This filter is only executed after the tests have completed while jacoco is
                 * generating the reports. Therefore, we cannot find the files list earlier. Although here were are
                 * finding the files list multiple times (for each file `f`), this is still acceptable because number of
                 * files are limited anyway.
                 */
                // This list will collect all the ".class" files from jacoco classes path folder
                List fs = new File(jacocoClassesPath).listFiles().findAll { it.name.endsWith(".class") }
                /*
                 * Next we need to keep only the file name (without path). Additionally, jacoco adds a session ID to the
                 * file name (e.g. something.class would look something like something.93153c79e479a123.class). What we
                 * are interested in here is only the "something.class" without the session ID.
                 * The `toString()` is also required otherwise the `.contains()` we do next will be comparing a G-String
                 * to a String and not match.
                 */
                fs = fs.collect { "${it.name.split('\\.')[0]}.class".toString() }
                // We only want to consider the current file `f` if its name does not already exist in `fs`.
                boolean keepFile = !fs.contains(f.name)
                /*
                 * XXX - it's enough to just return the value of `keepFile` without deletion but we do so to simplify
                 * the call to the jacoco plugin from Jenkins. If we don't do so, we'll need to do the filtering again
                 * from Jenkinsfile as well.
                 */
                if (!keepFile) {
                    f.delete()
                }
                return keepFile
            },
        ]
        classDirectories.from = allTrees
    }
}
```

**Important Note:** I'm not a gradle expert so the code above could be optimized but that's what I came up with.
For reference, below is my full `build.gradle` which I stitched from many different posts/comments over the time.
```
import groovy.json.JsonSlurper
import org.gradle.api.tasks.testing.logging.TestExceptionFormat
import org.gradle.api.tasks.testing.logging.TestLogEvent

group "jenkins-shared-library"
version "1.0-SNAPSHOT"

apply plugin: "groovy"
apply plugin: "idea"
apply plugin: "jacoco"

sourceCompatibility = 11
targetCompatibility = 11

idea {
    module {
        downloadJavadoc = true
        downloadSources = true
    }
}

sourceSets {
    main {
        groovy {
            srcDirs = ["vars", "src"]
        }
    }

    test {
        groovy {
            srcDirs = ["test"]
        }
    }
}

repositories {
    mavenCentral()
    maven { url "https://repo.jenkins-ci.org/releases/" }
    maven { url "https://repo.jenkins-ci.org/public/" }
}

dependencies {
    final String LINE = "\n" + "#" * 80 + "\n"

    // Jenkins core
    final String JENKINS_VERSION = System.getenv("JENKINS_VERSION") ?: "2.277.1"
    println "${LINE}Using Jenkins version ${JENKINS_VERSION}${LINE}"
    implementation group: "org.jenkins-ci.main", name: "jenkins-core", version: JENKINS_VERSION

    // Test frameworks
    implementation group: "org.codehaus.groovy", name: "groovy-all", version: "3.0.6"
    testImplementation group: "junit", name: "junit", version: "4.13.1"
    testImplementation group: "org.hamcrest", name: "hamcrest-all", version: "1.3"
    testImplementation group: "com.lesfurets", name: "jenkins-pipeline-unit", version: "1.9"
    testImplementation group: "org.mockito", name: "mockito-core", version: "3.5.13"

    // Jenkins plugins
    Map pluginsToLoad = [
        "com.cloudbees": [
            "groovy-cps",
        ],
        "io.jenkins.blueocean": [
            "blueocean-pipeline-api-impl",
            "blueocean-rest",
        ],
        "io.jenkins.plugins": [
            "code-coverage-api",
            "warnings-ng",
        ],
        "org.jenkins-ci.plugins": [
            "artifactory",
            "jacoco",
            "junit",
            "pipeline-build-step",
            "rich-text-publisher-plugin",
        ],
        "org.jenkins-ci.plugins.workflow": [
            "workflow-api",
            "workflow-basic-steps",
        ],
    ]

    // XXX - The following 2 lines are optional. In order to make sure we are testing against the right
    // plugins versions, extract those versions to a file from the pipeline and read it here.
    File pluginsFile = new File("plugins.json")
    Map pluginsVersions = pluginsFile.exists() ? new JsonSlurper().parse(pluginsFile) : [:]

    pluginsToLoad.each { String groupName, List groupPlugins ->
        groupPlugins.each { pluginName ->
            // The "+" sign indicates that the latest version should be used.
            String pluginVersion = "+"
            if (pluginName.contains(":")) {
                (pluginName, pluginVersion) = pluginName.split(":", 2)
            } else if (pluginsVersions[pluginName]) {
                pluginVersion = pluginsVersions[pluginName]
            }
            println "Including dependency '${groupName} -> ${pluginName}' with version '${pluginVersion}'."
            implementation group: groupName, name: pluginName, version: pluginVersion, ext: "jar"
        }
    }

    // TRICKY: The lib folder contains all other plugins *JAR* files if not found in Maven
    implementation fileTree(dir: "lib", include: ["*.jar"])
}

tasks.withType(Test) {
    // Run tests in parallel unless `PARALLEL_EXECUTION` environment variable is set to 0
    final int AVAILABLE_PROCESSORS = Runtime.runtime.availableProcessors()
    maxParallelForks = System.getenv("PARALLEL_EXECUTION") == "0" ? 1 : AVAILABLE_PROCESSORS.intdiv(2) ?: 1
    println maxParallelForks

    // We don't need the HTML report. We use junit xml only
    reports.html.enabled = false

    // A fancy test logger inspired by https://stackoverflow.com/a/36130467/3396919
    testLogging {
        // Set options for log level LIFECYCLE
        events TestLogEvent.FAILED,
            TestLogEvent.PASSED,
            TestLogEvent.SKIPPED,
            TestLogEvent.STANDARD_ERROR,
            TestLogEvent.STANDARD_OUT
        exceptionFormat TestExceptionFormat.FULL
        showExceptions true
        showCauses true
        showStackTraces true

        // set options for log level DEBUG and INFO
        debug {
            events TestLogEvent.STARTED,
                TestLogEvent.FAILED,
                TestLogEvent.PASSED,
                TestLogEvent.SKIPPED,
                TestLogEvent.STANDARD_ERROR,
                TestLogEvent.STANDARD_OUT
            exceptionFormat TestExceptionFormat.FULL
        }
        info.events = debug.events
        info.exceptionFormat = debug.exceptionFormat

        // Print the final report summary
        afterSuite { desc, result ->
            // Will match the outermost suite
            if (!desc.parent) {
                String output = "Results: ${result.resultType} (${result.testCount} tests, " +
                    "${result.successfulTestCount} passed, ${result.failedTestCount} failed, " +
                    "${result.skippedTestCount} skipped)"
                String startItem = "| ", endItem = " |"
                String linSep = "-" * (startItem.length() + output.length() + endItem.length())
                println("\n+${linSep}+\n${startItem} ${output} ${endItem}\n+${linSep}+")
            }
        }
    }
}

test {
    jacoco {
        classDumpDir = file("$buildDir/jacoco/classpathdumps")
    }

    finalizedBy jacocoTestReport
}

jacocoTestReport {
    dependsOn test

    afterEvaluate {
        /*
         * There are 2 location where the ".class" files exist:
         * 1) "$buildDir/jacoco/classpathdumps": This is where jacoco puts the class files which have been triggered by
         *    the tests.
         * 2) "$buildDir/classes/groovy/main": This is where gradle places all the class files during compilation
         *
         * It should be enough to just include the file tree from (1) above. However, this will result in not
         * considering any of the files which has not been tested.
         * In order to consider those as well, we add the class files from (2) as well. However, if we try to add the
         * same class multiple times, jacoco will complain. Therefore, we need to add all the files from (1), and the
         * files from (2) which are not in (1) already.
         */
        String jacocoClassesPath = "$buildDir/jacoco/classpathdumps"
        String gradleClassesPath = "$buildDir/classes/groovy/main"
        // The single "*" here indicates the "default" package which in our case is the "vars" dir
        List includesList = ["*", "com/company/**"]
        List excludesList = ["**/*Test*", "**/*Mock*"]

        List allTrees = [
            fileTree(dir: jacocoClassesPath, includes: includesList, excludes: excludesList),
            fileTree(dir: gradleClassesPath, includes: includesList, excludes: excludesList).filter { f ->
                /*
                 * In this filter, we want to remove all the files which were already covered by jacoco in the first
                 * `fileTree` above. This filter is only executed after the tests have completed while jacoco is
                 * generating the reports. Therefore, we cannot find the files list earlier. Although here were are
                 * finding the files list multiple times (for each file `f`), this is still acceptable because number of
                 * files are limited anyway.
                 */
                // This list will collect all the ".class" files from jacoco classes path folder
                @SuppressWarnings("JavaIoPackageAccess")
                List fs = new File(jacocoClassesPath).listFiles().findAll { it.name.endsWith(".class") }
                /*
                 * Next we need to keep only the file name (without path). Additionally, jacoco adds a session ID to the
                 * file name (e.g. something.class would look something like something.93153c79e479a993.class). What we
                 * are interested in here is only the "something.class" without the session ID.
                 * The `toString()` is also required otherwise the `.contains()` we do next will be comparing a G-String
                 * to a String and not match.
                 */
                fs = fs.collect { "${it.name.split('\\.')[0]}.class".toString() }
                // We only want to consider the current file `f` if its name does not already exist in `fs`.
                boolean keepFile = !fs.contains(f.name)
                /*
                 * XXX - it's enough to just return the value of `keepFile` without deletion but we do so to simplify
                 * the call to the jacoco plugin from Jenkins. If we don't do so, we'll need to do the filtering again
                 * from Jenkinsfile as well.
                 */
                if (!keepFile) {
                    f.delete()
                }
                return keepFile
            },
        ]
        classDirectories.from = allTrees
    }

    reports {
        xml.enabled true
        csv.enabled false
        html.enabled true
    }
}
```

Here is also part of my `Jenkinsfile` in my jenkins shared library repository which runs the tests and reports the coverage:
```
node("some_label_here") {
	// XXX- Following step is optional
	// In order to make sure we are testing against the right plugins versions, extract those versions
	// to a file and read it in build.gradle file.
	Map plugins = Jenkins.get().pluginManager.plugins.collectEntries { plugin ->
		return [(plugin.getShortName()): plugin.version]
	}
	writeJSON file: "plugins.json", json: plugins

	try {
		withEnv(["JENKINS_VERSION=${Jenkins.VERSION}"]) {
			sh "./gradlew clean test --warning-mode all"
		}
	} finally {
		junit "build/test-results/test/TEST-*.xml"
		jacoco(
			classPattern: "build/classes/groovy/main,build/jacoco/classpathdumps",
			inclusionPattern: "*,**/com/company/**",
			exclusionPattern: "**/*Test*,**/*Mock*",
			sourceInclusionPattern:"**/*.groovy",
			sourcePattern: "src/**,vars/**",
			runAlways: true
		)
	}
}
```

I hope this helps someone.

**Comment by deporcali** [comment]
> @deporcali This was no small feat! What's the "right" way to fix this?

@dblock This was just to enhance your workaround to overcome the limitations that it has

1. Only loaded classes will be dumped (i.e. classes not executed at all will not show-up in your report as not covered)
2. Report generation fails with `Can't add different class with same name` since this Jenkins unit testing framework will create several versions of the class at runtime. Jacoco dumps these class files with duplicate classes and therefore it throws this error

I do not know what is the "right" way to fix it. I would say to contribute to the [Jacoco project](https://github.com/jacoco/jacoco) and create a new release with the fix. The other thing I was considering is using another tool for code coverage. I have seen that the IntelliJ Code Coverage tool works as expected right out the box, so I was thinking into maybe looking at the [IntelliJ Code Coverage Agent](https://github.com/JetBrains/intellij-coverage) and see if it can be run locally

**Comment by mahmoud-ashi** [comment]
@deporcali take a look at my solution above. Regarding IntelliJ code coverage agent, I tried it some time ago inside the IDE itself and it works as you said however there are the following limitations that I remember:
- Cannot run tests in parallel
- Execution is much slower than with jacoco specially if collecting more traces
- The output report file is not standardized afaik so you can import it directly in other tools (e.g. jenkins plugins)

**Comment by deporcali** [comment]
> @deporcali take a look at my solution above. Regarding IntelliJ code coverage agent, I tried it some time ago inside the IDE itself and it works as you said however there are the following limitations that I remember:
> 
> * Cannot run tests in parallel
> * Execution is much slower than with jacoco specially if collecting more traces
> * The output report file is not standardized afaik so you can import it directly in other tools (e.g. jenkins plugins)

@mahmoud-ashi Your solution does not solves the issue of `Can't add different class with same name` since for example for the example below you only remove the session ID, but Jacoco can dump different ones of the same class depending on the number of modified classes at runtime

```bash
classdumpdir/foo.7bc27aq92d03dda5.class
classdumpdir/foo.aea79dlf0a9e3a59.class
```

Yeah I agree that the output of the IntelliJ code coverage agent is not easily consumable by other tools. I think there is some work regarding converting the output into something that other tools can understand.

**Comment by Pymann** [comment]
I can confirm the code examples by @deporcali  finalized by @mahmoud-ashi  work out of the box.

Very great thank you.

**Comment by Pymann** [comment]
@deporcali For me sourceSets.main.groovy.outputDir is not available?! Iam using gradle 8.4.

**Comment by Pymann** [comment]
with @deporcali  solution I get following error: 

`Could not get unknown property 'outputDir' for main Groovy source of type org.gradle.api.internal.tasks.DefaultGroovySourceDirectorySet.`
